oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/web/OAuth2TokenIntrospectionEndpointFilter.java

```diff
@@ -0,0 +1,450 @@
+			if (clientAuthentication == null || !clientAuthentication.getRegisteredClient().getId()
+					.equals(authorization.getRegisteredClientId())) {
+				throw new InvalidTokenException("Token does not correspond to authenticated client");
+			}
+
+			// we obtained the authorization from the token, no need to validate Optional here
+			AbstractOAuth2Token oauthToken = authorization.getTokens().getToken(token).get();
+			if (authorization.getTokens().getTokenMetadata(oauthToken).isInvalidated()) {
+				throw new InvalidTokenException("Token has been invalidated");
+			}
+
+			oauthToken = parseJwt(oauthToken);
+
+			validateToken(oauthToken);
+
+			TokenIntrospectionSuccessResponse.Builder builder = new TokenIntrospectionSuccessResponse.Builder(true);
+			builder.clientID(new ClientID(clientAuthentication.getRegisteredClient().getClientId()));
+			Optional.ofNullable(oauthToken.getIssuedAt()).map(Date::from).ifPresent(builder::issueTime);
+			Optional.ofNullable(oauthToken.getExpiresAt()).map(Date::from).ifPresent(builder::expirationTime);
+			TokenToIntrospectionResponseFieldsMapper.extractFromToken(oauthToken, builder);
+			return builder.build();
+		} catch (InvalidTokenException exception) {
+			TokenIntrospectionSuccessResponse.Builder builder = new TokenIntrospectionSuccessResponse.Builder(false);
+			return builder.build();
+		}
+	}
+
+	private void sendErrorResponse(HttpServletResponse response, OAuth2Error error) throws IOException {
+		ServletServerHttpResponse httpResponse = new ServletServerHttpResponse(response);
+		httpResponse.setStatusCode(HttpStatus.BAD_REQUEST);
+		this.errorHttpResponseConverter.write(error, null, httpResponse);
+	}
+
+	private Optional<OAuth2Authorization> findAuthorizationByToken(String token, Optional<TokenType> tokenTypeHint) {
+		List<TokenType> supportedTokenTypes = new ArrayList<>(
+				Arrays.asList(TokenType.ACCESS_TOKEN, TokenType.REFRESH_TOKEN, TokenType.ID_TOKEN));
+		tokenTypeHint.ifPresent(tType -> {
+			if (supportedTokenTypes.remove(tType))
+				supportedTokenTypes.add(0, tType);
+		});
+		for (TokenType tokenType : supportedTokenTypes) {
+			OAuth2Authorization authorization = this.authorizationService.findByToken(token, tokenType);
+			if (authorization != null) {
+				return Optional.of(authorization);
+			}
+		}
+		return Optional.empty();
+	}
+
+	private void sendTokenIntrospectionResponse(HttpServletResponse response,
+			TokenIntrospectionSuccessResponse tokenIntrospectionResponse) throws IOException {
+		ServletServerHttpResponse httpResponse = new ServletServerHttpResponse(response);
+		this.tokenIntrospectionHttpResponseConverter.write(tokenIntrospectionResponse, null, httpResponse);
+	}
+
+	private AbstractOAuth2Token parseJwt(AbstractOAuth2Token token) {
+		for (JwtDecoder jwtDecoder : this.jwtDecoders) {
+			try {
+				return jwtDecoder.decode(token.getTokenValue());
+			} catch (JwtException ex) {
+				// token might not be a JWT, or can't be processed with this decoder.
```

Comment:
Maybe we should go further here, and throw an _InvalidTokenException_  (which simply translates into an _{"active": false}_ response) if we detect the Token has been signed, but the signature is not valid? here we would treat it as an opaque token...
On the other hand, I guess if a JWT was tampered, then it won't work as an opaque token either, as it shouldn't be found within our stored tokens anyway. Let me know what you think.